File: android/mesh/src/main/java/io/godstone/mesh/diag/Diagnostics.kt
package io.godstone.mesh.diag

// ---------------------------------------------------------------------------
// T71 -- bounded local diagnostics, with no private telemetry (Android isle).
//
// The twin of tools/readiness/diagnostics.py and of the iOS Diagnostics.swift:
// the SAME vocabulary, the SAME laws, the SAME bounds. "Correctness under load
// needs measurable limits, but this offline app must not introduce tracking."
//
//   * OPT-IN (off by default; OFF counteth nothing);
//   * a CLOSED metric vocabulary -- an unknown name is REFUSED, because free names
//     are how a peer id or a body fragment reacheth a label;
//   * BOUNDED, saturating counters and non-negative gauges;
//   * MONOTONIC microsecond durations only -- no wall-clock timestamp is recorded;
//   * EPHEMERAL per-process relation ordinals -- a peer key never reacheth output;
//   * a BOUNDED drop-oldest ring whose supersessions are COUNTED;
//   * REDACTION BY REFUSAL: numbers and booleans only, everything else thrown.
//
// Nothing here transmitteth anything: the module carrieth no network type, and the
// shipping manifest removeth INTERNET.
// ---------------------------------------------------------------------------

/** The mode. OFF is the default, and OFF recordeth nothing at all. */
enum class DiagnosticsMode { OFF, ON }

/** A refused diagnostic. Thrown rather than silently dropped. */
class DiagnosticsRefusal(message: String) : IllegalArgumentException(message)

/** One rendered line: a metric name, a number, a monotonic duration. */
data class DiagnosticsLine(
    val metric: String,
    val value: Long,
    val durationMicros: Long,
    val relation: String,
) {
    fun render(): String =
        "metric=$metric value=$value dur_us=$durationMicros rel=$relation"
}

/**
 * The recorder. Every bound is a CONSTRUCTION: the ring cannot exceed its capacity,
 * a counter cannot exceed its ceiling, and the vocabulary cannot be widened by a
 * caller.
 */
class Diagnostics(
    val mode: DiagnosticsMode = DiagnosticsMode.OFF,
    val capacity: Int = RING_CAPACITY,
) {
    private val counters = LinkedHashMap<String, Long>()
    private val lines = ArrayDeque<DiagnosticsLine>()
    private var superseded = 0L
    private var refusals = 0L
    private var relationSeq = 0
    private val relationByKey = LinkedHashMap<String, String>()

    var currentMode: DiagnosticsMode = mode
        private set

    fun enable() { currentMode = DiagnosticsMode.ON }
    fun disable() { currentMode = DiagnosticsMode.OFF }
    val isOn: Boolean get() = currentMode == DiagnosticsMode.ON

    // ---- counters -------------------------------------------------------

    private fun checkMetric(metric: String?): String {
        if (metric == null || metric !in METRIC_NAMES) {
            throw DiagnosticsRefusal(
                "the metric name '$metric' is not in the vocabulary: free names are how private " +
                    "data reacheth a label")
        }
        return metric
    }

    /** REDACTION BY REFUSAL: only a number (or a boolean) may be a value. */
    private fun checkValue(value: Any?): Long {
        val asLong = when (value) {
            is Boolean -> if (value) 1L else 0L
            is Byte, is Short, is Int, is Long -> (value as Number).toLong()
            else -> throw DiagnosticsRefusal(
                "a diagnostic value must be a number; '${value?.let { it::class.simpleName }}' is " +
                    "exactly how a message body or a key fragment would enter the log")
        }
        return asLong.coerceIn(0L, COUNTER_CEILING)
    }

    fun count(metric: String?, value: Any? = 1L, durationMicros: Any? = 0L,
              relationKey: String? = null): Long {
        val name = checkMetric(metric)
        val amount = checkValue(value)
        if (!isOn) return 0L
        val duration = checkValue(durationMicros)
        val updated = minOf((counters[name] ?: 0L) + amount, COUNTER_CEILING)
        counters[name] = updated
        append(name, amount, duration, relationKey)
        return updated
    }

    fun gauge(metric: String?, value: Any?, relationKey: String? = null): Long {
        val name = checkMetric(metric)
        val amount = checkValue(value)
        if (!isOn) return 0L
        counters[name] = amount
        append(name, amount, 0L, relationKey)
        return amount
    }

    // ---- the ring -------------------------------------------------------

    private fun append(metric: String, value: Long, durationMicros: Long, relationKey: String?) {
        val line = DiagnosticsLine(metric, value, durationMicros, relation(relationKey))
        if (lines.size >= capacity) {
            lines.removeFirst()          // drop-oldest, and COUNT it
            superseded++
        }
        lines.addLast(line)
    }

    /** An EPHEMERAL relation id: per-process, opaque, never a node id. */
    fun relation(key: String? = null): String {
        if (key == null) {
            relationSeq++
            return "r$relationSeq"
        }
        val existing = relationByKey[key]
        if (existing != null) return existing
        relationSeq++
        val made = "r$relationSeq"
        relationByKey[key] = made
        // GS-DIAG-001: the MAP is bounded exactly as the ring is -- the audit reproduced a
        // ten-thousand-peer churn retaining every historic key in a second unbounded map.
        // An evicted key simply receiveth a FRESH ordinal if it returneth.
        while (relationByKey.size > capacity) {
            val eldest = relationByKey.keys.first()
            relationByKey.remove(eldest)
            superseded++
        }
        return made
    }

    val ringSize: Int get() = lines.size
    val isBounded: Boolean get() = ringSize <= capacity
    fun supersededCount(): Long = superseded
    fun refusalCount(): Long = refusals
    fun countersSnapshot(): Map<String, Long> = LinkedHashMap(counters)
    fun linesSnapshot(): List<DiagnosticsLine> = lines.toList()

    /** The whole output, as it would be written. THE ONLY ROAD OUT. */
    fun render(): String {
        val header = "diagnostics mode=${currentMode.name.lowercase()} lines=$ringSize " +
            "superseded=$superseded refusals=$refusals"
        return (listOf(header) + lines.map { it.render() }).joinToString("\n")
    }

    /** Everything this recorder knoweth vanisheth, and the ordinals restart. */
    fun reset() {
        counters.clear()
        lines.clear()
        superseded = 0
        refusals = 0
        relationSeq = 0
        relationByKey.clear()
    }

    /** Count one refusal (the court's own seam, so a refusal is VISIBLE). */
    fun noteRefusal() { refusals++ }

    companion object {
        /** The ring's bound: memory is bounded by construction, not by hope. */
        const val RING_CAPACITY: Int = 256

        /** A counter's ceiling: saturating, so a flood cannot overflow into nonsense. */
        const val COUNTER_CEILING: Long = 1L shl 40

        /** THE VOCABULARY. A metric may only be named from this set. */
        val METRIC_NAMES: Set<String> = linkedSetOf(
            "peers_seen", "peers_trusted", "frames_persisted", "frames_duplicate",
            "frames_forwarded", "frames_dropped_capacity", "queue_depth",
            "queue_superseded", "acks_admitted", "acks_refused", "archive_documents",
            "archive_bytes", "inference_started", "inference_cancelled",
            "inference_completed", "wipe_stages", "relations_opened", "relations_closed",
        )
    }
}

File: ios/Packages/GodstoneFoundation/Sources/GodstoneMesh/Diagnostics.swift
import Foundation

// ---------------------------------------------------------------------------
// T71 -- bounded local diagnostics, with no private telemetry (iOS isle).
//
// The twin of tools/readiness/diagnostics.py and of the Android Diagnostics.kt:
// the SAME vocabulary, the SAME laws, the SAME bounds.
//
//   * OPT-IN (off by default; OFF counteth nothing);
//   * a CLOSED metric vocabulary -- an unknown name is REFUSED;
//   * BOUNDED, saturating counters and non-negative gauges;
//   * MONOTONIC microsecond durations only -- no wall-clock date is recorded;
//   * EPHEMERAL per-process relation ordinals -- a peer key never reacheth output;
//   * a BOUNDED drop-oldest ring whose supersessions are COUNTED;
//   * REDACTION BY REFUSAL: numbers and booleans only, everything else thrown.
//
// Nothing here transmitteth anything: the module importeth no networking, and the
// shipping app carrieth no analytics.
// ---------------------------------------------------------------------------

/// The mode. OFF is the default, and OFF recordeth nothing at all.
public enum DiagnosticsMode: String, Sendable { case off, on }

/// A refused diagnostic. Thrown rather than silently dropped.
public struct DiagnosticsRefusal: Error, Equatable, CustomStringConvertible {
    public let reason: String
    public var description: String { reason }
}

/// One rendered line: a metric name, a number, a monotonic duration.
public struct DiagnosticsLine: Sendable, Equatable {
    public let metric: String
    public let value: Int64
    public let durationMicros: Int64
    public let relation: String

    public func render() -> String {
        "metric=\(metric) value=\(value) dur_us=\(durationMicros) rel=\(relation)"
    }
}

/// The recorder. Every bound is a CONSTRUCTION.
public final class Diagnostics: @unchecked Sendable {
    /// The ring's bound: memory is bounded by construction, not by hope.
    public static let ringCapacity = 256
    /// A counter's ceiling: saturating, so a flood cannot overflow into nonsense.
    public static let counterCeiling: Int64 = 1 << 40

    /// THE VOCABULARY. A metric may only be named from this set.
    public static let metricNames: [String] = [
        "peers_seen", "peers_trusted", "frames_persisted", "frames_duplicate",
        "frames_forwarded", "frames_dropped_capacity", "queue_depth",
        "queue_superseded", "acks_admitted", "acks_refused", "archive_documents",
        "archive_bytes", "inference_started", "inference_cancelled",
        "inference_completed", "wipe_stages", "relations_opened", "relations_closed",
    ]

    private let lock = NSLock()
    public let capacity: Int
    private var mode: DiagnosticsMode
    private var counters: [String: Int64] = [:]
    private var lines: [DiagnosticsLine] = []
    private var superseded: Int64 = 0
    private var refusals: Int64 = 0
    private var relationSeq = 0
    private var relationByKey: [String: String] = [:]

    public init(mode: DiagnosticsMode = .off, capacity: Int = Diagnostics.ringCapacity) {
        self.mode = mode
        self.capacity = capacity
    }

    public var currentMode: DiagnosticsMode { lock.lock(); defer { lock.unlock() }; return mode }
    public func enable() { lock.lock(); mode = .on; lock.unlock() }
    public func disable() { lock.lock(); mode = .off; lock.unlock() }
    public var isOn: Bool { currentMode == .on }

    // ---- counters -------------------------------------------------------

    private func checkedMetric(_ metric: String) throws -> String {
        guard Diagnostics.metricNames.contains(metric) else {
            throw DiagnosticsRefusal(reason: "the metric name '\(metric)' is not in the vocabulary: "
                                     + "free names are how private data reacheth a label")
        }
        return metric
    }

    /// REDACTION BY REFUSAL: only a number (or a boolean) may be a value.
    private func checkedValue(_ value: Any) throws -> Int64 {
        let raw: Int64
        if let flag = value as? Bool {
            raw = flag ? 1 : 0
        } else if let number = value as? Int64 {
            raw = number
        } else if let number = value as? Int {
            raw = Int64(number)
        } else {
            throw DiagnosticsRefusal(reason: "a diagnostic value must be a number; "
                                     + "'\(type(of: value))' is exactly how a message body or a key "
                                     + "fragment would enter the log")
        }
        return min(max(raw, 0), Diagnostics.counterCeiling)
    }

    @discardableResult
    public func count(_ metric: String, _ value: Any = Int64(1), durationMicros: Any = Int64(0),
                      relationKey: String? = nil) throws -> Int64 {
        let name = try checkedMetric(metric)
        let amount = try checkedValue(value)
        guard isOn else { return 0 }
        let duration = try checkedValue(durationMicros)
        lock.lock()
        let updated = min((counters[name] ?? 0) + amount, Diagnostics.counterCeiling)
        counters[name] = updated
        appendLocked(name, amount, duration, relationKey)
        lock.unlock()
        return updated
    }

    @discardableResult
    public func gauge(_ metric: String, _ value: Any, relationKey: String? = nil) throws -> Int64 {
        let name = try checkedMetric(metric)
        let amount = try checkedValue(value)
        guard isOn else { return 0 }
        lock.lock()
        counters[name] = amount
        appendLocked(name, amount, 0, relationKey)
        lock.unlock()
        return amount
    }

    // ---- the ring -------------------------------------------------------

    private func appendLocked(_ metric: String, _ value: Int64, _ durationMicros: Int64,
                              _ relationKey: String?) {
        let line = DiagnosticsLine(metric: metric, value: value,
                                   durationMicros: durationMicros,
                                   relation: relationLocked(relationKey))
        if lines.count >= capacity {
            lines.removeFirst()                     // drop-oldest, and COUNT it
            superseded += 1
        }
        lines.append(line)
    }

    /// An EPHEMERAL relation id: per-process, opaque, never a node id.
    public func relation(_ key: String? = nil) -> String {
        lock.lock(); defer { lock.unlock() }
        return relationLocked(key)
    }

    private func relationLocked(_ key: String?) -> String {
        guard let key else {
            relationSeq += 1
            return "r\(relationSeq)"
        }
        if let existing = relationByKey[key] { return existing }
        relationSeq += 1
        let made = "r\(relationSeq)"
        relationByKey[key] = made
        // GS-DIAG-001: the MAP is bounded exactly as the ring is -- the audit reproduced a
        // ten-thousand-peer churn retaining every historic key in a second unbounded map.
        // An evicted key simply receiveth a FRESH ordinal if it returneth.
        while relationByKey.count > capacity, let eldest = relationByKey.keys.first {
            relationByKey.removeValue(forKey: eldest)
            superseded += 1
        }
        return made
    }

    public var ringSize: Int { lock.lock(); defer { lock.unlock() }; return lines.count }
    public var isBounded: Bool { ringSize <= capacity }
    public var supersededCount: Int64 { lock.lock(); defer { lock.unlock() }; return superseded }
    public var refusalCount: Int64 { lock.lock(); defer { lock.unlock() }; return refusals }
    public var countersSnapshot: [String: Int64] { lock.lock(); defer { lock.unlock() }; return counters }
    public var linesSnapshot: [DiagnosticsLine] { lock.lock(); defer { lock.unlock() }; return lines }
    public func noteRefusal() { lock.lock(); refusals += 1; lock.unlock() }

    /// The whole output, as it would be written. THE ONLY ROAD OUT.
    public func render() -> String {
        lock.lock(); defer { lock.unlock() }
        let header = "diagnostics mode=\(mode.rawValue) lines=\(lines.count) "
            + "superseded=\(superseded) refusals=\(refusals)"
        return ([header] + lines.map { $0.render() }).joined(separator: "\n")
    }

    /// Everything this recorder knoweth vanisheth, and the ordinals restart.
    public func reset() {
        lock.lock()
        counters.removeAll(); lines.removeAll()
        superseded = 0; refusals = 0; relationSeq = 0; relationByKey.removeAll()
        lock.unlock()
    }
}
